x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java

```diff
@@ -0,0 +1,163 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.store;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.hash.MessageDigests;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * RoleReference is a handle to the actual role definitions (role descriptors).
+ * It has different sub-types depending on how the role descriptors should be resolved.
+ */
+public interface RoleReference {
+
+    /**
+     * The key for caching purpose. Callers can use this value to determine whether it should skip
+     * resolving the role descriptors and subsequently building the role.
+     */
+    RoleKey cacheKey();
+
+    /**
+     * Resolve concrete role descriptors for the roleReference.
+     */
+    void resolve(RoleReferenceResolver resolver, ActionListener<RolesRetrievalResult> listener);
+
+    /**
+     * Referencing a collection of role descriptors by their names
+     */
+    final class NamedRoleReference implements RoleReference {
+        private final String[] roleNames;
+
+        public NamedRoleReference(String[] roleNames) {
+            this.roleNames = roleNames;
+        }
+
+        public String[] getRoleNames() {
+            return roleNames;
+        }
+
+        @Override
+        public RoleKey cacheKey() {
+            if (roleNames.length == 0) {
+                return RoleKey.ROLE_KEY_EMPTY;
+            } else if (Arrays.asList(roleNames).contains(ReservedRolesStore.SUPERUSER_ROLE_DESCRIPTOR.getName())) {
+                return RoleKey.ROLE_KEY_SUPERUSER;
+            } else {
+                return new RoleKey(Set.copyOf(new HashSet<>(List.of(roleNames))), RoleKey.ROLES_STORE_SOURCE);
+            }
+        }
+
+        @Override
+        public void resolve(RoleReferenceResolver resolver, ActionListener<RolesRetrievalResult> listener) {
+            resolver.resolveNamedRoleReference(this, listener);
+        }
+    }
+
+    /**
+     * Referencing API Key role descriptors. Can be either the assigned (key) role descriptors or the limited-by (owner's) role descriptors
+     */
+    final class ApiKeyRoleReference implements RoleReference {
+
+        private final String apiKeyId;
+        private final BytesReference roleDescriptorsBytes;
+        private final String roleKeySource;
+
+        public ApiKeyRoleReference(String apiKeyId, BytesReference roleDescriptorsBytes, String roleKeySource) {
+            this.apiKeyId = apiKeyId;
+            this.roleDescriptorsBytes = roleDescriptorsBytes;
+            this.roleKeySource = roleKeySource;
+        }
+
+        @Override
+        public RoleKey cacheKey() {
+            final String roleDescriptorsHash = MessageDigests.toHexString(
+                MessageDigests.digest(roleDescriptorsBytes, MessageDigests.sha256())
+            );
```

Comment:
Should this key be memoized? If there's any chance we'd call the digest twice then that seems unnecessarily expensive.

Reply:
It is only called once by `CompositeRolesStore` and the result `RoleKey` is then passed down to other methods. So I don't think it is necessary memorize. That said, since we agree to change this method name to `id`. This gives it an impression that it could be called in other places (in future). So I added a simple class field to hold the result. Didn't use the MemorizedSupplier or volatile since they feel like overkill at this point.